bin/node-template/runtime/src/lib.rs

```diff
@@ -444,7 +444,7 @@ impl_runtime_apis! {
 	impl frame_benchmarking::Benchmark<Block> for Runtime {
 		fn dispatch_benchmark(
 			config: frame_benchmarking::BenchmarkConfig
-		) -> Result<Vec<frame_benchmarking::BenchmarkBatch>, sp_runtime::RuntimeString> {
+		) -> Result<Vec<frame_benchmarking::BenchmarkBatch>, Vec<u8>> {
```

Comment:
I kind of prefered the old type as it was more meaningful, looking at this type doesn't help understanding that it must return a `String` as a `Vec<u8>`.

Note that you could still build a function similar to `join_u8_sequences` on `RuntimeString` yourself.

The definition of RuntimeString is:
```rust

/// A string that wraps a `&'static str` in the runtime and `String`/`Vec<u8>` on decode.
#[derive(Eq, RuntimeDebug, Clone)]
pub enum RuntimeString {
	/// The borrowed mode that wraps a `&'static str`.
	Borrowed(&'static str),
	/// The owned mode that wraps a `String`.
	#[cfg(feature = "std")]
	Owned(String),
	/// The owned mode that wraps a `Vec<u8>`.
	#[cfg(not(feature = "std"))]
	Owned(Vec<u8>),
}
```
So you could maybe write:
```rust
fn runtime_string_from_parts(part: &[&'static str]) -> RuntimeString {
	// if std then concat into a String,
	// if not std then concat into a Vec,

	RuntimeString::Owned(final_value)
}
```

Reply:
I agree that it should be a `RuntimeString`. However, `RuntimeString` behaved weirdly in that it didn't seem to apply the `Owned(Vec<u8>)` for `#[cfg(not(feature = "std"))]` and always kept annoying me to give it a `String` which is not available with `no_std` features build. I wrote a question about it into the _Substrate technical chat_ (as written in the PR description) but got no reply. I paste the question text below, maybe you can make sense of it ;-)

> _ropottnik, 12.12.2020, substrate-technical-chat_
> hey guys! I'm trying to solve https://github.com/paritytech/substrate/issues/7386
as @Shawn Tabrizi#2662 pointed out on the issue, the right place to do it is inside the `add_benchmark!` macro inside https://github.com/paritytech/substrate/blob/master/frame/benchmarking/src/lib.rs#L1082 now one way or another, getting an improved error message would include concatenating strings inside a `map_err` statement. 
doing that also requires allocating memory which to me means that I need to return something owned. 

> In the error case I want to return a `RuntimeString` which has an `Owned(Vec<u8>)` variant for `no_std` features. 
> But here I'm stuck: When I go into the `node-template` and compile with `$ cargo build --features runtime-benchmarks`, it gives me the following error:

```
error[E0308]: mismatched types
   --> bin/node-template/runtime/src/lib.rs:466:4
    |
466 |             add_benchmark!(params, batches, frame_system, SystemBench::<Runtime>);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `std::string::String`, found struct `std::vec::Vec`
    |
    = note: expected struct `std::string::String`
               found struct `std::vec::Vec<u8>`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```
> Why would it want the `std::String` variant of the `RuntimeString` even though I'm not compiling with `std` features? 
> 
> Really stuck here, need help :wink:

Reply:
I think your compilation has triggered the `std` compilation of the node. (note that both std and not(std) are triggered when building the node).
My guess it that your code didn't returned a String in `std`, but a `Vec<u8>`


at the end your code should have something like this:
```rust
pub fn join_u8_sequences(seqs: &[&'static str]) -> sp_runtime::RuntimeString {
	#[cfg(feature = "std")]
	{
		sp_runtime::RuntimeString::Owned(seqs.iter().cloned().collect())
	}
	#[cfg(not(feature = "std"))]
	{
		sp_runtime::RuntimeString::Owned(seqs.iter().flat_map(|s| s.as_bytes()).cloned().collect())
	}
}
```